source/common/network/dns_resolver/dns_factory.cc

```diff
@@ -0,0 +1,70 @@
+#include "source/common/network/dns_resolver/dns_factory.h"
+
+namespace Envoy {
+namespace Network {
+
+// Create an empty c-ares DNS resolver typed config.
+void makeEmptyCaresDnsResolverConfig(
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
+  typed_dns_resolver_config.mutable_typed_config()->PackFrom(cares);
+  typed_dns_resolver_config.set_name(std::string(CaresDnsResolver));
+}
+
+// Create an empty apple DNS resolver typed config.
+void makeEmptyAppleDnsResolverConfig(
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  envoy::extensions::network::dns_resolver::apple::v3::AppleDnsResolverConfig apple;
+  typed_dns_resolver_config.mutable_typed_config()->PackFrom(apple);
+  typed_dns_resolver_config.set_name(std::string(AppleDnsResolver));
+}
+
+// Create an empty DNS resolver typed config based on build system and configuration.
+void makeEmptyDnsResolverConfig(
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  // If use apple API for DNS lookups, create an AppleDnsResolverConfig typed config.
+  if (checkUseAppleApiForDnsLookups(typed_dns_resolver_config)) {
+    return;
+  }
+  // Otherwise, create an CaresDnsResolverConfig typed config.
+  makeEmptyCaresDnsResolverConfig(typed_dns_resolver_config);
+}
+
+// If it is MacOS and the run time flag: envoy.restart_features.use_apple_api_for_dns_lookups
+// is enabled, create an AppleDnsResolverConfig typed config.
+bool checkUseAppleApiForDnsLookups(
+    envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) {
+  if ((Config::Utility::getAndCheckFactoryByName<Network::DnsResolverFactory>(
+           std::string(AppleDnsResolver), true) != nullptr) &&
```

Comment:
This is going to bite people in an unexpected way who do custom builds for apple since this is going to fail silently. Especially Envoy Mobile. cc @junr03 @goaway 

Can we make apple DNS a required extension on Apple/iOS builds?

Reply:
Yeah, let's make apple DNS as required extension as well, same as c-ares DNS.  This will make the behavior identical to the current behavior.

In Linux builds, apple DNS is not compiled in based on the BUILD file rules.  In MacOS builds,  it's built in.   People can toggle the run-time flag envoy.restart_features.use_apple_api_for_dns_lookups to choose c-ares DNS or apple DNS resolver library.